README.md
# python_rac_service

## Актуальный файл `service-3.py`


news.txt

IPv4 Statistics

  Packets Received                   = 80536321
  Received Header Errors             = 0
  Received Address Errors            = 6
  Datagrams Forwarded                = 0
  Unknown Protocols Received         = 14
  Received Packets Discarded         = 87202
  Received Packets Delivered         = 80545753
  Output Requests                    = 45380356
  Routing Discards                   = 0
  Discarded Output Packets           = 3053
  Output Packet No Route             = 518
  Reassembly Required                = 0
  Reassembly Successful              = 0
  Reassembly Failures                = 0
  Datagrams Successfully Fragmented  = 193
  Datagrams Failing Fragmentation    = 0
  Fragments Created                  = 523

IPv6 Statistics

  Packets Received                   = 4264434
  Received Header Errors             = 0
  Received Address Errors            = 48
  Datagrams Forwarded                = 0
  Unknown Protocols Received         = 1
  Received Packets Discarded         = 890
  Received Packets Delivered         = 4322250
  Output Requests                    = 874433
  Routing Discards                   = 0
  Discarded Output Packets           = 180
  Output Packet No Route             = 117
  Reassembly Required                = 0
  Reassembly Successful              = 0
  Reassembly Failures                = 0
  Datagrams Successfully Fragmented  = 0
  Datagrams Failing Fragmentation    = 0
  Fragments Created                  = 0

ICMPv4 Statistics

                            Received    Sent
  Messages                  1508        2457      
  Errors                    0           0         
  Destination Unreachable   1208        2087      
  Time Exceeded             23          0         
  Parameter Problems        0           0         
  Source Quenches           0           0         
  Redirects                 0           0         
  Echo Replies              269         8         
  Echos                     8           362       
  Timestamps                0           0         
  Timestamp Replies         0           0         
  Address Masks             0           0         
  Address Mask Replies      0           0         
  Router Solicitations      0           0         
  Router Advertisements     0           0         

ICMPv6 Statistics

                            Received    Sent
  Messages                  8202        8374      
  Errors                    0           0         
  Destination Unreachable   142         545       
  Packet Too Big            0           0         
  Time Exceeded             0           0         
  Parameter Problems        0           1         
  Echos                     0           404       
  Echo Replies              293         0         
  MLD Queries               0           0         
  MLD Reports               0           0         
  MLD Dones                 0           0         
  Router Solicitations      0           168       
  Router Advertisements     1221        0         
  Neighbor Solicitations    4578        2397      
  Neighbor Advertisements   1968        4859      
  Redirects                 0           0         
  Router Renumberings       0           0         

TCP Statistics for IPv4

  Active Opens                        = 40726
  Passive Opens                       = 1141
  Failed Connection Attempts          = 13508
  Reset Connections                   = 3750
  Current Connections                 = 85
  Segments Received                   = 72027424
  Segments Sent                       = 39393498
  Segments Retransmitted              = 0

TCP Statistics for IPv6

  Active Opens                        = 7205
  Passive Opens                       = 517
  Failed Connection Attempts          = 4696
  Reset Connections                   = 567
  Current Connections                 = 0
  Segments Received                   = 419097
  Segments Sent                       = 191472
  Segments Retransmitted              = 0

UDP Statistics for IPv4

  Datagrams Received    = 8922490
  No Ports              = 86757
  Receive Errors        = 73
  Datagrams Sent        = 4395092

UDP Statistics for IPv6

  Datagrams Received    = 3921869
  No Ports              = 814
  Receive Errors        = 0
  Datagrams Sent        = 619922


service-1.py
import json
import os


FILENAME = "command.json"
WORKDIR = os.getcwd()

os.system("chcp 65001")

with open(FILENAME, "r") as file:
    command_json = json.load(file)

for arg in command_json["args"]:
    command = str(command_json["command"]) + " " + arg + f" > new{arg}.txt"
    os.system(command)
# os.system("netstat -s > new.txt")


def lines_to_keys(lines):
    line_dict = {}

    for line in lines:
        key, values = line.split("=")
        line_dict[key.strip()] = values.strip()

    return line_dict


def without_equal(lines):
    keys = []
    values = []

    for line in lines:
        splited = line.split()
        length = len(splited)
        keys.append(" ".join(splited[: length - 2]))
        values.append([splited[length - 2], splited[length - 1]])

    return {
        key: {"Recieved": values[0], "Sent": values[1]}
        for key, values in zip(keys, values)
    }


def lines_to_dict_s(lines):
    keys = []
    values = []
    tmp1 = []  # without =
    tmp2 = []  # for =

    for line in lines:
        if "Received" in line and "Sent" in line:
            continue
        elif "=" in line:
            tmp2.append(line)
        elif "Pv" not in line:
            tmp1.append(line)
        else:
            if len(tmp2) != 0:
                values.append(lines_to_keys(tmp2))
                tmp2.clear()
            if len(tmp1) != 0:
                values.append(without_equal(tmp1))
                tmp1.clear()
            keys.append(line.strip())
    values.append(lines_to_keys(tmp2))
    tmp2.clear()

    return {key: value for key, value in zip(keys, values)}


def for_s():
    with open("new-s.txt", "r") as f:
        content = f.readlines()
    # os.remove(WORKDIR + r"\\new-s.txt")

    content = [line.strip() for line in content if line != "\n"]
    lines = [lines_to_dict_s(content)]

    with open(WORKDIR + r"\\output-s.json", "w") as file:
        json.dump(lines, file, indent=4)

    return 0


def lines_to_dict_t(lines):
    keys = lines[0].split("  ")
    keys = [key.strip() for key in keys if key != ""]
    values = [line.split() for line in lines[1:]]
    return [
        {key: value for key, value in zip(keys, values[i])}
        for i in range(len(values) - 1)
    ]


def for_t():
    with open("new-t.txt", "r") as f:
        content = f.readlines()[3:]
    # os.remove(WORKDIR + r"\\new-t.txt")

    content = [line.strip() for line in content if line != "\n"]
    lines = [lines_to_dict_t(content)]

    with open(WORKDIR + r"\\output-t.json", "w") as file:
        json.dump(lines, file, indent=4)

    return 0


def main():
    if "-s" in command_json["args"]:
        for_s()

    if "-t" in command_json["args"]:
        for_t()


if __name__ == "__main__":
    main()


service-3.py
from flask import Flask, request, jsonify
import os


FILENAME = "data.txt"
app = Flask(__name__)


def main(mode, cmd, opts=None):
    os.system("chcp 65001")

    if opts is None:
        os.system(f"./rac {mode} {cmd} > {FILENAME}")
    else:
        os.system(f"./rac {mode} {cmd} {opts} > {FILENAME}")

    keys, values = [], []
    with open(FILENAME, "r") as f:
        text = f.readlines()
    for line in text:
        key, value = line.split(":")
        keys.append(key.strip())
        values.append(value.strip())

    return {key: value for key, value in zip(keys, values)}


@app.route("/", methods=["GET", "POST"])
def output():
    help = """
    PLEASE USE POST REQUEST!
    Fields:
    mode=%%mode_name%%
    command=%%command_name%%
    options=%%list_of_options%% (may be empty)

    Example:
    curl -X POST -H "Content-Type: application/json"
    -d '{"mode": "cluster", "command": "list"}'
    https://127.0.0.1:3000/
    """

    if request.method != "POST":
        return help

    request_data = request.get_json()
    mode = request_data["mode"]
    if not mode or mode == "":
        return "Error: Invalid Mode"
    if "command" in request_data.keys():
        command = request_data["command"]
    else:
        return "Error: Invalid Command"
    if "options" in request_data.keys():
        options = request_data["options"]
    else:
        options = None

    return jsonify(main(mode, command, options))


if __name__ == "__main__":
    app.debug = True
    app.run(host="127.0.0.1", port=3000)


service-2.py
from flask import Flask, request, jsonify
import json
import os

app = Flask(__name__)

WORKDIR = os.getcwd()


def lines_to_keys(lines):
    line_dict = {}

    for line in lines:
        key, values = line.split("=")
        line_dict[key.strip()] = values.strip()

    return line_dict


def without_equal(lines):
    keys = []
    values = []

    for line in lines:
        splited = line.split()
        length = len(splited)
        keys.append(" ".join(splited[: length - 2]))
        values.append([splited[length - 2], splited[length - 1]])

    return {
        key: {"Recieved": values[0], "Sent": values[1]}
        for key, values in zip(keys, values)
    }


def lines_to_dict_s(lines):
    keys = []
    values = []
    tmp1 = []  # without =
    tmp2 = []  # for =

    for line in lines:
        if "Received" in line and "Sent" in line:
            continue
        elif "=" in line:
            tmp2.append(line)
        elif "Pv" not in line:
            tmp1.append(line)
        else:
            if len(tmp2) != 0:
                values.append(lines_to_keys(tmp2))
                tmp2.clear()
            if len(tmp1) != 0:
                values.append(without_equal(tmp1))
                tmp1.clear()
            keys.append(line.strip())
    values.append(lines_to_keys(tmp2))
    tmp2.clear()

    return {key: value for key, value in zip(keys, values)}


def for_s():
    with open("new-s.txt", "r") as f:
        content = f.readlines()
    os.remove(WORKDIR + r"\\new-s.txt")

    content = [line.strip() for line in content if line != "\n"]
    # with open(WORKDIR + r"\\output-s.json", "w") as file:
    #     json.dump(lines, file, indent=4)

    return [lines_to_dict_s(content)]


def lines_to_dict_t(lines):
    keys = lines[0].split("  ")
    keys = [key.strip() for key in keys if key != ""]
    values = [line.split() for line in lines[1:]]
    return [
        {key: value for key, value in zip(keys, values[i])}
        for i in range(len(values) - 1)
    ]


def for_t():
    with open("new-t.txt", "r") as f:
        content = f.readlines()[3:]
    os.remove(WORKDIR + r"\\new-t.txt")

    content = [line.strip() for line in content if line != "\n"]
    # with open(WORKDIR + r"\\output-t.json", "w") as file:
    #     json.dump(lines, file, indent=4)

    return lines_to_dict_t(content)


def main(cmd, args):
    os.system("chcp 65001")
    args = args.split(",")

    for arg in args:
        command = f"{cmd} -{arg} > new-{arg}.txt"
        os.system(command)
    # os.system("netstat -s > new.txt")

    output_json = []

    if "s" in args:
        output_json.append(for_s())

    if "t" in args:
        output_json.append(for_t())

    # with open(WORKDIR + r"\\output.json", "w") as file:
    #     json.dump(out_s + out_t, file, indent=2)

    return output_json


@app.route("/", methods=["GET"])
def output():
    command = request.args.get("command")
    if not command or command == "":
        return "Error: Invalid Parameters"
    args = request.args.get("args")

    return jsonify(main(command, args))


if __name__ == "__main__":
    app.debug = True
    app.run(host="127.0.0.1", port=3000)
